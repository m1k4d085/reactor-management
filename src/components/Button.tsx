import { ComponentProps } from "react";
import { StyleType, ThemeType, generateButtonStyle } from "../util";

interface ButtonProps
  extends Omit<ComponentProps<"button">, "className" | "color"> {
  styleType?: StyleType;
  themeType?: ThemeType;
}

function Button({
  styleType = "solid",
  themeType = "primary",
  ...buttonProps
}: ButtonProps) {
  return (
    <button
      className={generateButtonStyle(themeType, styleType)}
      {...buttonProps}
    />
  );
}
export default Button;
