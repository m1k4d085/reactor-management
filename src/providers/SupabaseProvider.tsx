import { PropsWithChildren } from "react";
import { Supabase, SupabaseFunctions } from "../contexts/Supabase";

interface SupabaseProviderProps extends PropsWithChildren {
  supabase: SupabaseFunctions;
}

function SupabaseProvider({ supabase, children }: SupabaseProviderProps) {
  return <Supabase.Provider value={supabase}>{children}</Supabase.Provider>;
}
export default SupabaseProvider;
