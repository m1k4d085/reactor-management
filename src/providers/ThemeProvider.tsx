import { PropsWithChildren, useEffect, useState } from "react";
import { Theme } from "../contexts/Theme";
import { ModeType, ThemeType } from "../util";
import { themeBrowser } from "../theme/theme";

interface ThemeProviderProps extends PropsWithChildren {}

function ThemeProvider({ children }: ThemeProviderProps) {
  const [modeType, setModeType] = useState<ModeType>(themeBrowser.getMode());
  const [themeType, setThemeType] = useState<ThemeType>(
    themeBrowser.getTheme()
  );

  useEffect(() => {
    themeBrowser.setMode(modeType);
  }, [modeType]);

  useEffect(() => {
    themeBrowser.setTheme(themeType);
  }, [themeType]);

  return (
    <Theme.Provider
      value={{
        modeType,
        themeType,
        setMode(mode) {
          setModeType(mode);
        },
        setTheme(theme) {
          setThemeType(theme);
        },
        switchMode() {
          const newMode = modeType === "dark" ? "light" : "dark";
          setModeType(newMode);
          return newMode;
        },
      }}
    >
      {children}
    </Theme.Provider>
  );
}
export default ThemeProvider;
