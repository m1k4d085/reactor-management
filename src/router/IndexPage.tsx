function IndexPage() {
  return (
    <div id="index-page" className="h-full">
        <div className="flex flex-col items-center justify-center h-full">
            <span className="text-5xl font-bold">Reactor</span>
            <span className="text-5xl font-bold">Management</span>
        </div>
    </div>
  )
}
export default IndexPage