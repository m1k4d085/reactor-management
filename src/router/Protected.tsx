import { PropsWithChildren } from "react";
import { useAppSelector } from "../state";
import { Navigate, To } from "react-router-dom";

interface ProtectedProps extends PropsWithChildren {
  redirectTo: To;
}

function Protected({ redirectTo, children }: ProtectedProps) {
  const auth = useAppSelector(({ auth }) => auth);
  console.log(auth.loggedIn)
  if (auth.loggedIn) return children;
  else return <Navigate to={redirectTo} />;
}
export default Protected;
