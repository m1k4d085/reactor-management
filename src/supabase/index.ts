import {
  SignInWithPasswordCredentials,
  createClient,
} from "@supabase/supabase-js";
import { SupabaseFunctions } from "../contexts/Supabase";
import { Tables, TablesInsert } from "./database.types";

const client = createClient(
  import.meta.env.VITE_SUPABASE_URL,
  import.meta.env.VITE_SUPABASE_KEY
);

async function signIn(credentials: SignInWithPasswordCredentials) {
  const { data, error } = await client.auth.signInWithPassword(credentials);
  if (error) throw error;
  return data;
}

async function signOut() {
  const { error } = await client.auth.signOut();
  if (error) throw error;
}

async function recoverSession() {
  const {
    data: { session },
    error: sessionError,
  } = await client.auth.getSession();
  if (sessionError) throw sessionError;

  if (session?.expires_at && session.expires_at < Date.now()) {
    const { data, error: refreshError } = await client.auth.refreshSession();
    if (refreshError) throw refreshError;

    return data;
  }

  if (!session) throw new Error("Impossibile recuperare la sessione");

  // const {
  //   data: { user },
  //   error: userError,
  // } = await client.auth.getUser();
  // if (userError) throw userError;

  return { user: session.user, session };
}

export const supabase: SupabaseFunctions = {
  signIn,
  signOut,
  recoverSession,
  products: {
    async get(query?: Partial<Tables<"products">>) {
      let select = client.from("products").select();
      if (query) {
        let field: keyof typeof query;
        for (field in query) {
          const value = query[field];
          if (value) select = select.eq(field, value);
        }
      }
      const { data, error } = await select;
      if (error) throw error;

      return data;
    },
    async getOne({ id }: { id: string }) {
      const { data, error } = await client
        .from("products")
        .select()
        .eq("id", id)
        .single();
      if (error) throw error;

      return data;
    },
    async insert({ entity }: { entity: TablesInsert<"products"> }) {
      const { data, error } = await client
        .from("products")
        .insert(entity)
        .select()
        .single();
      if (error) throw error;

      return data;
    },
    async update({
      id,
      entity,
    }: {
      id: string;
      entity: TablesInsert<"products">;
    }) {
      const { data, error } = await client
        .from("products")
        .update(entity)
        .eq("id", id)
        .select()
        .single();
      if (error) throw error;

      return data;
    },
    async remove({ id }: { id: string }) {
      const { error } = await client.from("products").delete().eq("id", id);
      if (error) throw error;
    },
  },
  media: {
    async get(query?: Partial<Tables<"media">>) {
      let select = client.from("media").select();
      if (query) {
        let field: keyof typeof query;
        for (field in query) {
          const value = query[field];
          if (value) select = select.eq(field, value);
        }
      }
      const { data, error } = await select;
      if (error) throw error;

      return data;
    },
    async getOne({ id }: { id: string }) {
      const { data, error } = await client
        .from("media")
        .select()
        .eq("id", id)
        .single();
      if (error) throw error;

      return data;
    },
    async insert({ entity }: { entity: TablesInsert<"media"> }) {
      const { data, error } = await client
        .from("media")
        .insert(entity)
        .select()
        .single();
      if (error) throw error;

      return data;
    },
    async update({
      id,
      entity,
    }: {
      id: string;
      entity: TablesInsert<"media">;
    }) {
      const { data, error } = await client
        .from("media")
        .update(entity)
        .eq("id", id)
        .select()
        .single();
      if (error) throw error;

      return data;
    },
    async remove({ id }: { id: string }) {
      const { error } = await client.from("media").delete().eq("id", id);
      if (error) throw error;
    },
  },
  customers: {
    async get(query?: Partial<Tables<"customers">>) {
      let select = client.from("customers").select();
      if (query) {
        let field: keyof typeof query;
        for (field in query) {
          const value = query[field];
          if (value) select = select.eq(field, value);
        }
      }
      const { data, error } = await select;
      if (error) throw error;

      return data;
    },
    async getOne({ id }: { id: string }) {
      const { data, error } = await client
        .from("customers")
        .select()
        .eq("id", id)
        .single();
      if (error) throw error;

      return data;
    },
    async insert({ entity }: { entity: TablesInsert<"customers"> }) {
      const { data, error } = await client
        .from("customers")
        .insert(entity)
        .select()
        .single();
      if (error) throw error;

      return data;
    },
    async update({
      id,
      entity,
    }: {
      id: string;
      entity: TablesInsert<"customers">;
    }) {
      const { data, error } = await client
        .from("customers")
        .update(entity)
        .eq("id", id)
        .select()
        .single();
      if (error) throw error;

      return data;
    },
    async remove({ id }: { id: string }) {
      const { error } = await client.from("customers").delete().eq("id", id);
      if (error) throw error;
    },
  },
  addresses: {
    async get(query?: Partial<Tables<"addresses">>) {
      let select = client.from("addresses").select();
      if (query) {
        let field: keyof typeof query;
        for (field in query) {
          const value = query[field];
          if (value) select = select.eq(field, value);
        }
      }
      const { data, error } = await select;
      if (error) throw error;

      return data;
    },
    async getOne({ id }: { id: string }) {
      const { data, error } = await client
        .from("addresses")
        .select()
        .eq("id", id)
        .single();
      if (error) throw error;

      return data;
    },
    async insert({ entity }: { entity: TablesInsert<"addresses"> }) {
      const { data, error } = await client
        .from("addresses")
        .insert(entity)
        .select()
        .single();
      if (error) throw error;

      return data;
    },
    async update({
      id,
      entity,
    }: {
      id: string;
      entity: TablesInsert<"addresses">;
    }) {
      const { data, error } = await client
        .from("addresses")
        .update(entity)
        .eq("id", id)
        .select()
        .single();
      if (error) throw error;

      return data;
    },
    async remove({ id }: { id: string }) {
      const { error } = await client.from("addresses").delete().eq("id", id);
      if (error) throw error;
    },
  },
  orders: {
    async get(query?: Partial<Tables<"orders">>) {
      let select = client.from("orders").select();
      if (query) {
        let field: keyof typeof query;
        for (field in query) {
          const value = query[field];
          if (value) select = select.eq(field, value);
        }
      }
      const { data, error } = await select;
      if (error) throw error;

      return data;
    },
    async getOne({ id }: { id: string }) {
      const { data, error } = await client
        .from("orders")
        .select()
        .eq("id", id)
        .single();
      if (error) throw error;

      return data;
    },
    async insert({ entity }: { entity: TablesInsert<"orders"> }) {
      const { data, error } = await client
        .from("orders")
        .insert(entity)
        .select()
        .single();
      if (error) throw error;

      return data;
    },
    async update({
      id,
      entity,
    }: {
      id: string;
      entity: TablesInsert<"orders">;
    }) {
      const { data, error } = await client
        .from("orders")
        .update(entity)
        .eq("id", id)
        .select()
        .single();
      if (error) throw error;

      return data;
    },
    async remove({ id }: { id: string }) {
      const { error } = await client.from("orders").delete().eq("id", id);
      if (error) throw error;
    },
  },
  orderLines: {
    async get(query?: Partial<Tables<"order_lines">>) {
      let select = client.from("orders").select();
      if (query) {
        let field: keyof typeof query;
        for (field in query) {
          const value = query[field];
          if (value) select = select.eq(field, value);
        }
      }
      const { data, error } = await select;
      if (error) throw error;

      return data;
    },
    async getOne({ id }: { id: string }) {
      const { data, error } = await client
        .from("order_lines")
        .select()
        .eq("id", id)
        .single();
      if (error) throw error;

      return data;
    },
    async insert({ entity }: { entity: TablesInsert<"order_lines"> }) {
      const { data, error } = await client
        .from("order_lines")
        .insert(entity)
        .select()
        .single();
      if (error) throw error;

      return data;
    },
    async update({
      id,
      entity,
    }: {
      id: string;
      entity: TablesInsert<"order_lines">;
    }) {
      const { data, error } = await client
        .from("order_lines")
        .update(entity)
        .eq("id", id)
        .select()
        .single();
      if (error) throw error;

      return data;
    },
    async remove({ id }: { id: string }) {
      const { error } = await client.from("order_lines").delete().eq("id", id);
      if (error) throw error;
    },
  },
};
