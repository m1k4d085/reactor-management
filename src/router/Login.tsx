import { useNavigate } from "react-router-dom";
import { useSupabase } from "../hooks/useSupabase";
import { useAppDispatch, useAppSelector } from "../state";
import { Form, Formik } from "formik";
import TextInput from "../components/TextInput";
import Button from "../components/Button";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSignIn } from "@fortawesome/free-solid-svg-icons";
import { signIn } from "../state/authSlice";
import { useEffect } from "react";

interface SignInForm {
  email: string;
  password: string;
}

function Login() {
  const supabase = useSupabase();
  const auth = useAppSelector(({ auth }) => auth);
  const dispatch = useAppDispatch();
  const navigate = useNavigate();

  async function submit(values: SignInForm) {
    const data = await supabase?.signIn(values);
    if (data) dispatch(signIn(data));
  }

  useEffect(() => {
    if (auth.loggedIn) navigate("/");
  }, [auth.loggedIn, navigate]);

  return (
    <div className="w-screen h-screen bg-light dark:bg-dark text-dark dark:text-light flex justify-center items-center">
      <div className="bg-light dark:bg-dark shadow-lg dark:shadow-lg dark:shadow-light px-10 py-12 z-30">
        <h1 className="text-center text-2xl font-bold">Login</h1>
        <Formik<SignInForm>
          initialValues={{
            email: "",
            password: "",
          }}
          onSubmit={submit}
        >
          {({ handleChange }) => (
            <Form>
              <TextInput
                label="Email"
                type="email"
                name="email"
                onChange={handleChange}
              />
              <TextInput
                label="Password"
                type="password"
                name="password"
                onChange={handleChange}
              />
              <div className="flex justify-center mt-5">
                <Button type="submit">
                  <div className="flex gap-2 items-center">
                    <FontAwesomeIcon icon={faSignIn} size="2xl" />
                    <span>Login</span>
                  </div>
                </Button>
              </div>
            </Form>
          )}
        </Formik>
      </div>
    </div>
  );
}
export default Login;
