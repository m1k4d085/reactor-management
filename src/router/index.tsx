import { createBrowserRouter } from "react-router-dom";
import Root from "./Root";
import Login from "./Login";
import IndexPage from "./IndexPage";
import Storybook from "../pages/Storybook";
import Storycolor from "../pages/Storycolor";
import { fakeLoader } from "../util";
import Protected from "./Protected";
import Test from "../pages/Test";
import Customers from "../pages/customers/Customers";
import FormCustomer from "../pages/customers/FormCustomer";
import Media from "../pages/products/Media";
import Customer from "../pages/customers/Customer";

export interface CustomerParams extends Record<string, string | undefined> {
  customerId: string | undefined;
}

export const router = createBrowserRouter([
  {
    path: "/",
    element: <Root />,
    children: [
      {
        index: true,
        element: <IndexPage />,
        loader: fakeLoader,
      },
      {
        path: "/media",
        element: (
          <Protected redirectTo="/login">
            <Media />
          </Protected>
        ),
        loader: fakeLoader,
      },
      {
        path: "/customers",
        element: (
          <Protected redirectTo="/login">
            <Customers />
          </Protected>
        ),
        loader: fakeLoader,
      },
      {
        path: "/customers/add",
        element: (
          <Protected redirectTo="/login">
            <FormCustomer />
          </Protected>
        ),
        loader: fakeLoader,
      },
      {
        path: "/customers/:customerId",
        element: (
          <Protected redirectTo="/login">
            <Customer />
          </Protected>
        ),
        loader: fakeLoader,
      },
      {
        path: "/customers/:customerId/edit",
        element: (
          <Protected redirectTo="/login">
            <FormCustomer />
          </Protected>
        ),
        loader: fakeLoader,
      },
      {
        path: "/storybook",
        element: <Storybook />,
        loader: fakeLoader,
      },
      {
        path: "/storycolor",
        element: <Storycolor />,
        loader: fakeLoader,
      },
      {
        path: "/test",
        element: (
          <Protected redirectTo="/login">
            <Test />
          </Protected>
        ),
        loader: fakeLoader,
      },
    ],
  },
  {
    path: "/login",
    element: <Login />,
  },
]);
