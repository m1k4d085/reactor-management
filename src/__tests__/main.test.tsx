import { render, renderHook, waitFor } from "@testing-library/react";
import { test, expect } from "vitest";
// import { sleep } from "../util";
import { QueryClient, QueryClientProvider } from "react-query";
import { Provider as ReduxProvider } from "react-redux";
import { RouterProvider } from "react-router-dom";
import SupabaseProvider from "../providers/SupabaseProvider";
import ThemeProvider from "../providers/ThemeProvider";
import { queryClient } from "../query";
import { router } from "../router";
import { store } from "../state";
import { supabase } from "../supabase";
import Button from "../components/Button";
import useCustomers from "../hooks/Customers/useSelect";
import { FC, PropsWithChildren } from "react";

test("add", () => {
  expect(1 + 1).toBe(2); //jest
});

test("add type", () => {
  // const myNumber = Number(2);
  // expectTypeOf(myNumber).toBeBoolean;
});

// describe.concurrent("prova", () => {
//   test("serial", async () => {
//     await sleep(2000);
//   });

//   test.concurrent("concurrent 1", async () => {
//     await sleep(2000);
//   });
//   test.concurrent("concurrent 2", async () => {
//     await sleep(2000);
//   });
// });

// test.concurrent("single concurrent 1", async () => {
//   await sleep(2000);
// });
// test.concurrent("single concurrent 2", async () => {
//   await sleep(2000);
// });

test("render app", async () => {
  render(
    <ThemeProvider>
      <SupabaseProvider supabase={supabase}>
        <QueryClientProvider client={queryClient}>
          <ReduxProvider store={store}>
            <RouterProvider router={router} />
          </ReduxProvider>
        </QueryClientProvider>
      </SupabaseProvider>
    </ThemeProvider>
  );
});

test("button", async () => {
  const result = render(<Button />);
  expect(result).toMatchSnapshot();
});

test("useCustomer", async () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        staleTime: Infinity,
      },
    },
  });

  const wrapper: FC<PropsWithChildren> = ({ children }) => (
    <SupabaseProvider supabase={supabase}>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </SupabaseProvider>
  );

  const { result: customers } = renderHook(useCustomers, { wrapper });

  await waitFor(() => expect(customers.current[1].status).toBe("success"));

  expect(customers.current[1].data).toHaveLength(2);
});
