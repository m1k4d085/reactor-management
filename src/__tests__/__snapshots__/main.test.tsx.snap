// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`button 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <button
        class="px-4 py-3 rounded-t-md rounded-b-md disabled:opacity-50 disabled:cursor-not-allowed bg-primary-500 text-light hover:bg-primary-700 active:bg-primary-800 disabled:bg-primary-200 disabled:text-dark dark:bg-primary-500 dark:text-dark dark:hover:bg-primary-300 dark:active:bg-primary-200 dark:disabled:bg-primary-800 dark:disabled:text-light"
      />
    </div>
  </body>,
  "container": <div>
    <button
      class="px-4 py-3 rounded-t-md rounded-b-md disabled:opacity-50 disabled:cursor-not-allowed bg-primary-500 text-light hover:bg-primary-700 active:bg-primary-800 disabled:bg-primary-200 disabled:text-dark dark:bg-primary-500 dark:text-dark dark:hover:bg-primary-300 dark:active:bg-primary-200 dark:disabled:bg-primary-800 dark:disabled:text-light"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
